Add Compositor Effects (Post-Processing) demo

BastiaanOlij opened the pull request:
This demo project shows two fairly simple examples on how to use compositor effects to create post effects.

The example that is in use applies a template shader into which user code can be injected, this user code is stored in a property on the compositor effect.
The second example uses a `glsl` file containing the full compute shader, which is compiled at initialisation and then used. 

![image](https://github.com/godotengine/godot-demo-projects/assets/1945449/73cf224f-2299-458f-bffe-6e71f1d9d65c)

I'm working on a documentation page for the Godot documentation that will provide more detail and uses this demo project as its base. 

This replaces #942 as a simpler example.
This demo is based on the project build during this stream: https://www.youtube.com/watch?v=KLmahauWeV4

BastiaanOlij committed:
Compute demo for compositor effect

AThousandShips commented on compute/post_shader/post_process_grayscale.gd:
```diff
@@ -0,0 +1,86 @@
+# Compile our shader at initialisation.
+func _initialize_compute():
+	rd = RenderingServer.get_rendering_device()
+	if !rd:
```
```suggestion
	if not rd:
```

AThousandShips commented on compute/post_shader/post_process_grayscale.gd:
```diff
@@ -0,0 +1,86 @@
+
+
+# Called by the rendering thread every frame.
+func _render_callback(p_effect_callback_type, p_render_data):
```
```suggestion
func _render_callback(effect_callback_type, render_data):
```

BastiaanOlij replied:
This cant be done @AThousandShips , it will cause name clashes with member variables.

AThousandShips commented on compute/post_shader/post_process_shader.gd:
```diff
@@ -0,0 +1,158 @@
+
+# Check if our shader has changed and needs to be recompiled.
+func _check_shader() -> bool:
+	if !rd:
```
```suggestion
	if not rd:
```

AThousandShips commented on compute/post_shader/post_process_shader.gd:
```diff
@@ -0,0 +1,158 @@
+
+
+# Called by the rendering thread every frame.
+func _render_callback(p_effect_callback_type, p_render_data):
```
```suggestion
func _render_callback(effect_callback_type, render_data):
```

BastiaanOlij replied:
Same here, we can't make this chance, removing the prefix will cause name clashes and giving parameters a non-sensible name, doesn't make sense.

AThousandShips commented on compute/post_shader/post_process_shader.gd:
```diff
@@ -0,0 +1,158 @@
+		return false
+
+	shader = rd.shader_create_from_spirv(shader_spirv)
+	if !shader.is_valid():
```
```suggestion
	if not shader.is_valid():
```

AThousandShips commented on compute/post_shader/post_process_shader.gd:
```diff
@@ -0,0 +1,158 @@
+		return false
+
+	pipeline = rd.compute_pipeline_create(shader)
+	if !pipeline.is_valid():
+		return false
+
+	return true
```
```suggestion
	return pipeline.is_valid()
```

BastiaanOlij replied:
Hmm, I'll make this change but this breaks a pattern I try to follow. In this case we have a process, at each step of the process we check for failure and exit. When all steps are done, you exit true for success.
Doing this results in every step following that pattern, while now the last step looks out of place.
And doing this means that someone who comes around later and adds another step to the process, simply continues the pattern instead of having to wonder why the pipeline being valid is the success criteria.

AThousandShips reviewed (approved):
Style looks good!

clayjohn commented on compute/post_shader/README.md:
```diff
@@ -0,0 +1,39 @@
+
+Renderer: Forward+
+
+> Note: this demo requires Godot 4.2 or later
```
```suggestion
> Note: this demo requires Godot 4.3 or later
```

BastiaanOlij replied:
oops :)

clayjohn reviewed (approved):
Tested locally and it works great. The example seems good to me, it is a good level of complexity for a demo

BastiaanOlij commented:
Just a ping, it would be nice if we could merge this now that the documentation pages referring to this demo have been merged.

Calinou commented:
New demos should have type hints following https://github.com/godotengine/godot-demo-projects/pull/1063. I'll look into pushing a commit on top of this PR today.

Calinou committed:
Use a different shader-based effect, make both effects toggleable at runtime

The shader-based effect now inverts colors to be distinguishable from
the other grayscale effect.

This also adds type hints to scripts and improves visuals (MSAA,
anisotropic filtering, lossless compression on pattern texture).

Calinou commented:
Thanks!
